src/Analysis/Engine/Impl/Indexing/SymbolIndexWalker.cs

```diff
@@ -0,0 +1,267 @@
+            var children = _stack.Exit();
+
+            _stack.AddSymbol(new HierarchicalSymbol(
+                node.Name,
+                SymbolKind.Class,
+                node.GetSpan(_ast),
+                node.NameExpression.GetSpan(_ast),
+                children,
+                FunctionKind.Class
+            ));
+
+            return false;
+        }
+
+        public override bool Walk(FunctionDefinition node) {
+            _stack.Enter(SymbolKind.Function);
+            foreach (var p in node.Parameters) {
+                AddVarSymbol(p.NameExpression);
+            }
+            node.Body?.Walk(this);
+            var children = _stack.Exit();
+
+            var span = node.GetSpan(_ast);
+
+            var ds = new HierarchicalSymbol(
+                node.Name,
+                SymbolKind.Function,
+                span,
+                node.IsLambda ? span : node.NameExpression.GetSpan(_ast),
+                children,
+                FunctionKind.Function
+            );
+
+            if (_stack.Parent == SymbolKind.Class) {
+                switch (ds.Name) {
+                    case "__init__":
+                        ds.Kind = SymbolKind.Constructor;
+                        break;
+                    case var name when DoubleUnderscore.IsMatch(name):
+                        ds.Kind = SymbolKind.Operator;
+                        break;
+                    default:
+                        ds.Kind = SymbolKind.Method;
+
+                        if (node.Decorators != null) {
+                            foreach (var dec in node.Decorators.Decorators) {
+                                var maybeKind = DecoratorExpressionToKind(dec);
+                                if (maybeKind.HasValue) {
+                                    ds.Kind = maybeKind.Value.Item1;
+                                    ds._functionKind = maybeKind.Value.Item2;
+                                    break;
+                                }
+                            }
+                        }
+
+                        break;
+                }
+            }
+
+            _stack.AddSymbol(ds);
+
+            return false;
+        }
+
+        public override bool Walk(ImportStatement node) {
+            foreach (var imp in node.Names.Zip(node.AsNames, (name, asName) => (name, asName))) {
+                if (imp.asName == null) {
+                    var span = imp.name.GetSpan(_ast);
+                    _stack.AddSymbol(new HierarchicalSymbol(imp.name.MakeString(), SymbolKind.Module, span));
+                } else {
+                    var span = imp.asName.GetSpan(_ast);
+                    _stack.AddSymbol(new HierarchicalSymbol(imp.asName.Name, SymbolKind.Module, span));
+                }
+            }
+
+            return false;
+        }
+
+        public override bool Walk(FromImportStatement node) {
+            foreach (var imp in node.Names.Zip(node.AsNames, (name, asName) => (name, asName))) {
```

Comment:
`foreach (var (name, asName) in node.Zip...)`

Reply:
Actually:
```
foreach (var nameNode node.Names.Zip(node.AsNames, (name, asName) => asName ?? name))) {
    var span = nameNode.GetSpan(_ast);
    _stack.AddSymbol(new HierarchicalSymbol(nameNode.Name, SymbolKind.Module, span));
}

Reply:
Ooo, yes, way better.

Reply:
Agh, unfortunately doesn't help the other `Zip`.

Reply:
Actually, I can just generate the `span` early.

Reply:
Maybe not. There's not a super clean way to do the one above unless I write:

```csharp
foreach (var (name, span) in node.Names.Zip(node.AsNames,
	(name, asName) => (asName?.Name ?? name.MakeString(), ((asName as Node) ?? (name as Node)).GetSpan(_ast)))) {
	_stack.AddSymbol(new HierarchicalSymbol(name, SymbolKind.Module, span));
}
```

Reply:
```cs
foreach (var (nameNode, nameString) in node.Names.Zip(node.AsNames, (name, asName) => asName != null ? (asName, asName.Name) : (name, name.MakeString()) {
    var span = nameNode.GetSpan(_ast);
    _stack.AddSymbol(new HierarchicalSymbol(nameString, SymbolKind.Module, span));
}
```

Reply:
That doesn't quite work, since the compiler can't figure out how to combine `asName` and `name`, but I can do `asName as Node` or `name as Node` and have it figure it out.

Reply:
i.e.:
```csharp
foreach (var (nameNode, nameString) in node.Names.Zip(node.AsNames, (name, asName) => asName != null ? (asName, asName.Name) : (name as Node, name.MakeString()) {
	var span = nameNode.GetSpan(_ast);
	_stack.AddSymbol(new HierarchicalSymbol(nameString, SymbolKind.Module, span));
}
```
Up to you if you think that looks better than the plain if statement.

Reply:
To me it looks better

Reply:
`(Node)name` or `node as Name`? This cast can never fail, so `as` is technically overkill.

Reply:
I casted, but let me know if you want `as` instead for looks.

Reply:
`(Node)name` is better

Reply:
Great, it's done already then.